dot/core/service.go

```diff
@@ -454,258 +382,37 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
-			}
-		}
-		// body
-		if (blockRequest.RequestedData&2)>>1 == 1 {
-			retData, err := s.blockState.GetBlockBody(hash)
-			if err == nil && retData != nil {
-				blockData.Body = retData.AsOptional()
-			}
-		}
-		// receipt
-		if (blockRequest.RequestedData&4)>>2 == 1 {
-			retData, err := s.blockState.GetReceipt(hash)
-			if err == nil && retData != nil {
-				blockData.Receipt = optional.NewBytes(true, retData)
-			}
-		}
-		// message queue
-		if (blockRequest.RequestedData&8)>>3 == 1 {
-			retData, err := s.blockState.GetMessageQueue(hash)
-			if err == nil && retData != nil {
-				blockData.MessageQueue = optional.NewBytes(true, retData)
-			}
-		}
-		// justification
-		if (blockRequest.RequestedData&16)>>4 == 1 {
-			retData, err := s.blockState.GetJustification(hash)
-			if err == nil && retData != nil {
-				blockData.Justification = optional.NewBytes(true, retData)
-			}
-		}
-
-		responseData = append(responseData, blockData)
-	}
-
-	return &network.BlockResponseMessage{
-		ID:        blockRequest.ID,
-		BlockData: responseData,
-	}, nil
-}
-
-// ProcessBlockResponseMessage attempts to validate and add the block to the
-// chain by calling `core_execute_block`. Valid blocks are stored in the block
-// database to become part of the canonical chain.
-func (s *Service) ProcessBlockResponseMessage(msg network.Message) error {
-	log.Debug("[core] received BlockResponseMessage")
-	s.respOut <- msg.(*network.BlockResponseMessage)
-
-	return s.checkForRuntimeChanges()
-}
-
-// checkForRuntimeChanges checks if changes to the runtime code have occurred; if so, load the new runtime
-func (s *Service) checkForRuntimeChanges() error {
-	currentCodeHash, err := s.storageState.LoadCodeHash()
-	if err != nil {
-		return err
-	}
-
-	if !bytes.Equal(currentCodeHash[:], s.codeHash[:]) {
-		code, err := s.storageState.LoadCode()
-		if err != nil {
-			return err
-		}
-
-		s.rt.Stop()
+func (s *Service) handleBabeSession() {
+	for {
+		// wait for BABE epoch to complete
+		<-s.epochDone
 
-		s.rt, err = runtime.NewRuntime(code, s.storageState, s.keys)
+		// verify first block was set during current epoch
+		err := s.verifyFirstBlock()
 		if err != nil {
-			return err
+			log.Error(
+				"[core] first block in epoch not set",
+				"block", s.firstBlock,
+				"error", err,
+			)
 		}
 
-		// kill babe session, handleBabeSession will reload it with the new runtime
-		if s.isBabeAuthority {
-			err = s.safeBabeKill()
-			if err != nil {
-				return err
-			}
-		}
-	}
-
-	return nil
-}
-
-// ProcessTransactionMessage validates each transaction in the message and
-// adds valid transactions to the transaction queue of the BABE session
-func (s *Service) ProcessTransactionMessage(msg network.Message) error {
-
-	// get transactions from message extrinsics
-	txs := msg.(*network.TransactionMessage).Extrinsics
-
-	for _, tx := range txs {
-		tx := tx // pin
+		// update current epoch number
+		s.epochNumber = s.epochNumber + 1
```

Comment:
does this work with syncing?

Reply:
related to https://github.com/ChainSafe/gossamer/pull/703#discussion_r398772061

Reply:
revised in latest commit